abm_sketch.py

```diff
@@ -35,182 +37,330 @@
+
             self.success_lens.append(len(self.path))
             self.path = []
 
-        self.points[recipient].receive_task(task, sender)
+
+    def calculate_award(self, task, path, entity):
+        """
+        Returns the amount awarded to <entity> after <task> has been
+        routed through <path>.
+        """
+        k = k = float(len(path))
+        return (task.value / k) + ((k - path.index(entity)) * 5)
+
 
     def set_connections(self):
-        self.connectivity_matrix = get_connectivity_matrix(self.points)
+        """Initializes each Entity's adjacency list."""
+
+        self.make_connectivity_matrix()
         for index, point in enumerate(self.points):
             point.set_adjacencies(self.connectivity_matrix[index])
 
-    def display(self):
-        display(self.points, self.connectivity_matrix, show=self.show)
+
+    def make_connectivity_matrix(self):
+        """
+        Computes the connectivity matrix of this Population. Each point is
+        connected to each other within a radius.
+        """
+
+        if self.connectivity_matrix:
+            return
+
+        points_arr = np.array([[p.x, p.y] for p in self.points])
+        distance_mat = euclidean_distances(points_arr, points_arr)
+
+        # Every point p will be connected to each other point whose distance
+        # to p is less than a cut-off value. This value is computed as the
+        # mean of {min_nonzero(dist_mat(p)) | p is a point}, times a factor
+        min_nonzero = lambda r: min(r[r > 0])
+
+        # apply_along_axis(f, axis=1, arr) applies f to each row
+        min_neighbor_distances = np.apply_along_axis(min_nonzero, axis=1, arr=distance_mat)
+
+        factor = 2.2
+        neighbor_cutoff = np.mean(min_neighbor_distances) * factor
+        connectivity_matrix = distance_mat < neighbor_cutoff
+
+        self.connectivity_matrix = connectivity_matrix
+
+
+    def display(self, current=None, target=None):
+        """
+        Plots the state of the task. If <show> = False, doesn't plot
+        anything and the simulation can run faster.
+        """
+
+        if not self.show:
+            return
+
+        # Scatter plot of points, color coded by class
+        pts = self.points
+        size = 35
+        for cluster, color in cluster_colors.iteritems():
+            class_points = [x for x in pts if x.cluster == cluster]
+            plt.scatter([p.x for p in class_points], [p.y for p in class_points], c=color, s=size)
+
+        # Draw the connections
+        if self.connectivity_matrix is not None:
+            for start_ix, connections in enumerate(self.connectivity_matrix):
+                for connect_ix, connected in enumerate(connections):
+                    if connected and connect_ix != start_ix:
+                        plt.plot(*zip(
+                            (pts[start_ix].x, pts[start_ix].y),
+                            (pts[connect_ix].x, pts[connect_ix].y)),
```

Comment:
For the new network generation mechanism I have working now, the location of the point is irrelevant. So I was looking into modifying the locations of each point every time and draw points closer as their likelihood to share a message increases.

But leaving them in place and using the edges sounds easier.